### PR TITLE
feat(dataset): make dataset providers plugins

### PR DESCRIPTION
**Pull request by olevski**
As I am working through #2971 one of the subtasks is to make the dataset providers plugins.

I isolated the changes for this and thought it would be a good idea to submit this PR separately from the rest.

Also wanted to get a feeling of whether my way of going about this is acceptable. I decided to not try to separate `ProviderApi` and the similar importer and exporter abstract classes and try to move some portion of them in the domain_model. But if you think this is necessary then I am happy to do it. 

For now the plugins simply provide the defintion of a specific ProviderApi. And the provider factory is now using pluggy to list all the providers rather than importing what is only locally available.

### CODE REVIEW

File: renku/core/dataset/providers/dataverse.py [hunk 1]
```diff
@@ -580,3 +582,12 @@ def make_file_url(file_id, base_url):
     args_dict = {"persistentId": file_id}
     url_parts[4] = urllib.parse.urlencode(args_dict)
     return urllib.parse.urlunparse(url_parts)
+
+
+class DataverseProviderPlugin(IDatasetProviderPlugin):
```
Comment (Panaetius) [inline renku/core/dataset/providers/dataverse.py:587]: Could we add this as a mixin on e.g. `DataverseProvider` directly?

Like 
```
class DataverseProvider(ProviderApi, IDatasetProviderPlugin):
[...]
    @hookimpl
    @classmethod  # not sure on the correct order of decorators
    def dataset_provider(cls) -> "Type[ProviderApi]":
        """The defintion of the provider."""
        return cls
[...]
```
Same for the other providers.

**Review by Panaetius — APPROVED** [review]
Looks all great. I left on comment on how we could simplify it by adding the method on the provider directly (then we don't need extra classes), but that doesn't change anything functionality-wise, so feel free to ignore.

File: renku/core/dataset/providers/dataverse.py (hunk 1 above)
Comment (olevski) [reply renku/core/dataset/providers/dataverse.py:587]: Oh cool. A class method will do exactly what I was looking for. Don't know why I did not think of this.

I will also fix the mypy and style problems. Just wanted to see if this approach was acceptable or not.

**Review by Panaetius — DISMISSED** [review]
Thank you, glad it worked with the classmethod :slightly_smiling_face:

**Comment by olevski** [comment]
@Panaetius can I get a re-approve here? There were merge conflicts I had to resolve and that cleared the approval.

**Review by Panaetius — APPROVED** [review]

**Pull request merged**